### PR TITLE
Add weights on edges to pagerank implementation

### PR DESCRIPTION
**Pull request by yinghaodai**
#132

### CODE REVIEW

**Comment by yinghaodai** [comment]
@vandenheuvel Please have a look especially at the changed delta in the test with two nodes. Previously, it assumed that the rank of both nodes should be between 0.4 and 0.6, with high probability. However, now that we are using a directed graph in which the weights are taken into account, it can be that one of the nodes sends way more data to the other than vice versa, and thereby having a greater page rank. Therefore, the original tests failed multiple times, which caused me to change the delta from 0.1 to 0.2. Do you agree with this?

File: Tribler/Test/Community/Multichain/statistics/test_page_rank.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ def create_graph(self, nr_of_nodes, nr_of_edges):
         :param nr_of_nodes: amount of nodes in the random graph
         :param nr_of_edges: amount of edges in the random graph
         """
-        self.graph = gnm_random_graph(nr_of_nodes, nr_of_edges)
+        self.graph = gnm_random_graph(nr_of_nodes, nr_of_edges, directed=True)
```
Comment (vandenheuvel) [inline Tribler/Test/Community/Multichain/statistics/test_page_rank.py]: Please build the same, original graph without this method to keep the test the same.

File: Tribler/community/multichain/statistics/page_rank.py [hunk 2]
```diff
@@ -40,17 +40,16 @@ def __init__(self, graph, walks_per_node=2, stop_probability=0.1):
 
         self.new_nodes = set()
 
-    def add_edge(self, source, destination):
+    def add_edge(self, source, destination, weight=1):
```
Comment (vandenheuvel) [inline Tribler/community/multichain/statistics/page_rank.py]: Please remove this `=1`.

File: Tribler/community/multichain/statistics/page_rank.py [hunk 3]
```diff
@@ -106,7 +105,14 @@ def _walk(self, start):
             neighbors = self.graph.neighbors(next_node)
             if len(neighbors) == 0:
                 break
-            next_node = choice(neighbors)
+            weighted_neighbors = []
+            for neighbor in neighbors:
```
Comment (vandenheuvel) [inline Tribler/community/multichain/statistics/page_rank.py]: Please rewrite this without lists.

File: Tribler/Test/Community/Multichain/output/annotations.txt [hunk 4]
```diff
@@ -0,0 +1,250 @@
+annotation start end
```
Comment (ghabbenjansen) [inline Tribler/Test/Community/Multichain/output/annotations.txt]: Please remove this file

File: Tribler/Test/Core/Modules/RestApi/output/annotations.txt [hunk 5]
```diff
@@ -0,0 +1,15 @@
+annotation start end
```
Comment (ghabbenjansen) [inline Tribler/Test/Core/Modules/RestApi/output/annotations.txt]: Also remove this file

File: Tribler/Test/Community/Multichain/statistics/test_page_rank.py [hunk 6]
```diff
@@ -13,14 +14,16 @@ class TestIncrementalPageRank(TestCase):
     As it is a Monte Carlo method, integration tests are often probabilistic in nature.
     That is, they might fail on rare occasions.
     """
-    def create_graph(self, nr_of_nodes, nr_of_edges, nr_of_times=1):
+    def create_graph(self, nr_of_nodes, nr_of_edges, nr_of_times=1, directed=True):
```
Comment (vandenheuvel) [inline Tribler/Test/Community/Multichain/statistics/test_page_rank.py]: I think we can require the graph to be directed, so you can remove this optional argument.

File: Tribler/Test/Community/Multichain/statistics/test_page_rank.py [hunk 7]
```diff
@@ -13,14 +14,16 @@ class TestIncrementalPageRank(TestCase):
     As it is a Monte Carlo method, integration tests are often probabilistic in nature.
     That is, they might fail on rare occasions.
```
Comment (vandenheuvel) [inline Tribler/Test/Community/Multichain/statistics/test_page_rank.py]: Please mention that the graph is directed in the documentation.

**Comment by vandenheuvel** [comment]
retest this please

**Pull request merged**